### ISSUE TITLE
Workflow instance read only variables

### ISSUE THREAD

**Issue by sergeylebed** [camunda/camunda#3219]
Variables that are passed from Workflow element to element are changeable by these elements, so it might easy happen, especially if you have a lot of different elements developed by different developers, that some variable names may conflict and be accidentally overridden.

I would like to have read only properties (variables) that are passed as an additional argument when you create an instance of a workflow. These Instance Properties should be workflow wide and accessible by all elements and be available in the job parameter in a worker, as job.InstanceProperties()

So, they are like workflow wide headers (https://github.com/zeebe-io/zeebe/issues/3218) but they are created not during workflow design time, but during workflow instance instantiation.

Examples:
#1. The payload that is processed by workflow elements might have a big size and it might be expensive to keep it in Zeebe as well as it might have performance implication. Instead I want to store the payload on external storage and provide just a reference to this data as Zeebe workflow instance property, I must be sure that this reference could not be changed by any worker, so the same data is accessible by all workers.

#2. I want to run a common workflow for a specific user. So, I want to provide a user id to this workflow. I must be sure that none of worker is able to changes this user id.

**Comment by jwulf** [comment]
What happens when a worker tries to complete a job, overriding a read-only variable. Does it throw over gRPC in the client?

**Comment by sergeylebed** [comment]
There's no such thing as read only variables, so nothing happens. It is a suggestion to have such variables that could be provided when you create and instance of the task.

I think it is better to call such read only variables as Instance Properties to clearly distinguish them from variables that can be modified by workers.

**Comment by themichaelscott** [comment]
I am looking for something similar. I was wondering if headers could be used as a solution, but it seems they can only be set at the time you are modeling your workflow, not at instance creation.  Is there a way to set a header when an instance is created? This would be a viable alternative for me if it is possible. I don't see a way to do it though.

**Comment by saig0** [comment]
> Is there a way to set a header when an instance is created?

No. Currently, custom task headers are statically defined on the design-time of the workflow.

**Comment by npepinpe** [comment]
Seems reasonable, will bring it up to the product team :+1:

**Comment by jzhongsun** [comment]
any update on the issue?

**Comment by npepinpe** [comment]
Unfortunately not - this is an interesting feature, but it's not currently aligned with any of our goals, so it's on ice until we have some resources for it.
